core/src/main/java/com/linecorp/armeria/common/DefaultHttpData.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.util.Objects;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.internal.common.ByteArrayBytes;
+import com.linecorp.armeria.internal.common.ByteBufBytes;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.util.ResourceLeakHint;
+
+/**
+ * A {@code byte[]}-based {@link HttpData}.
+ */
+final class DefaultHttpData implements HttpData, ResourceLeakHint {
+
+    static final DefaultHttpData empty = new DefaultHttpData(ByteArrayBytes.empty(), false);
+    static final DefaultHttpData emptyEos = new DefaultHttpData(ByteArrayBytes.empty(), true);
+
+    static HttpData of(byte[] data) {
+        return new DefaultHttpData(ByteArrayBytes.of(data));
+    }
+
+    static HttpData of(ByteBuf data, boolean pooled) {
+        return new DefaultHttpData(ByteBufBytes.of(data, pooled));
+    }
```

Comment:
These delegations look flexible but create an additional object.
`HttpData` is a massively-used type for transporting data.
So I slightly prefer an inherence style rather than composition although it can cause to create many subtypes.
```java
final class ByteArrayHttpData extends ByteArrayBytes implements HttpData, ResourceLeakHint {
    ...
}

final class ByteBufHttpData extends ByteBufBytes implements HttpData, ResourceLeakHint {
    ...
}
```

Reply:
`WebSocketFrame`, which is not `HttpData`, will use `ByteArrayBytes` and `ByteBufBytes`. 
We are not creating a heavy object but a light obejct which just has `Bytes`, so shouldn't it be fine?

Reply:
> WebSocketFrame, which is not `HttpData`, will use `ByteArrayBytes` and `ByteBufBytes`

I didn't mean to remove `ByteArrayBytes` and `ByteBufBytes`.
`WebSocketFrame` can still use them but `final` modifier for the classes would be removed to `ByteArrayHttpData` and `ByteBufHttpData` extends them.

Reply:
I just changed my previous comment just before you left it. Yeah, that sounds reasonable.